File: articles/present-bias-immediate-reward/cpp/present_bias_choice.cpp
#include <cmath>
#include <iostream>

double discounted_delayed_value(double beta, double delta, double reward, int delay) {
    return beta * std::pow(delta, delay) * reward;
}

bool choose_delayed_reward(double beta, double delta, double delayed_reward, int delay, double immediate_reward, double commitment_cost) {
    double delayed_value = discounted_delayed_value(beta, delta, delayed_reward, delay);
    double immediate_value = immediate_reward - commitment_cost;
    return delayed_value >= immediate_value;
}

int main() {
    std::cout << "Synthetic delayed choice under present bias: "
              << choose_delayed_reward(0.72, 0.97, 300.0, 12, 160.0, 70.0)
              << std::endl;
    return 0;
}

File: articles/present-bias-immediate-reward/go/present_bias_choice.go
package main

import (
	"fmt"
	"math"
)

func discountedDelayedValue(beta, delta, reward float64, delay int) float64 {
	return beta * math.Pow(delta, float64(delay)) * reward
}

func chooseDelayedReward(beta, delta, delayedReward float64, delay int, immediateReward, commitmentCost float64) bool {
	delayedValue := discountedDelayedValue(beta, delta, delayedReward, delay)
	immediateValue := immediateReward - commitmentCost
	return delayedValue >= immediateValue
}

func main() {
	fmt.Printf("Synthetic delayed choice under present bias: %v\n", chooseDelayedReward(0.72, 0.97, 300.0, 12, 160.0, 70.0))
}
